File: runtime/cpp/titan-ast-runtime-lib/BacktrackingBottomUpBranch.cpp
//
// Created by tian wei jun on 2022/12/4 0004.
//

#include "BacktrackingBottomUpBranch.h"
#include <unordered_set>

BacktrackingBottomUpBranch::BacktrackingBottomUpBranch()
    : reducingSymbols(std::list<ReducingSymbol *>()) {}

BacktrackingBottomUpBranch::~BacktrackingBottomUpBranch() {
  for (auto reducingSymbol : reducingSymbols) {
    delete reducingSymbol;
    reducingSymbol = nullptr;
  }
  reducingSymbols.clear();
}

BacktrackingBottomUpBranch *
BacktrackingBottomUpBranch::cloneForAstAutomata() const {
  auto *bottomUpBranch = new BacktrackingBottomUpBranch();

  for (auto reducingSymbol : reducingSymbols) {
    bottomUpBranch->reducingSymbols.push_back(
        reducingSymbol->cloneForAstAutomata());
  }
  return bottomUpBranch;
}

// for BacktrackingBottomUpAstAutomata.triedBottomUpBranchs(set)
bool BacktrackingBottomUpBranch::compare(
    const BacktrackingBottomUpBranch *o) const {
  if (this->reducingSymbols.back()->endIndexOfToken != o->reducingSymbols.back()->endIndexOfToken) {
    return this->reducingSymbols.back()->endIndexOfToken < o->reducingSymbols.back()->endIndexOfToken;
  }

  if (this->reducingSymbols.size() != o->reducingSymbols.size()) {
    return this->reducingSymbols.size() < o->reducingSymbols.size();
  }
  auto thisReducingSymbolsIt = this->reducingSymbols.rbegin();
  auto oReducingSymbolsIt = o->reducingSymbols.rbegin();
  while (thisReducingSymbolsIt != this->reducingSymbols.rend()) {
    ReducingSymbol *thisReducingSymbol = *thisReducingSymbolsIt;
    ReducingSymbol *oReducingSymbol = *oReducingSymbolsIt;

    auto compare = thisReducingSymbol->compare(oReducingSymbol);

    if (0 != compare) {
      return compare < 0;
    }
    thisReducingSymbolsIt++;
    oReducingSymbolsIt++;
  }
  return false;
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/BacktrackingBottomUpBranch.java
package titan.ast.runtime;

import java.util.LinkedList;
import java.util.ListIterator;

/**
 * 自顶向上归约分支.
 *
 * @author tian wei jun
 */
public class BacktrackingBottomUpBranch
    implements Cloneable, Comparable<BacktrackingBottomUpBranch> {

  LinkedList<ReducingSymbol> reducingSymbols = new LinkedList<>();

  public BacktrackingBottomUpBranch cloneForAstAutomata() {
    BacktrackingBottomUpBranch cloner = new BacktrackingBottomUpBranch();
    cloner.reducingSymbols = new LinkedList<>();
    for (ReducingSymbol reducingSymbol : this.reducingSymbols) {
      cloner.reducingSymbols.addLast(reducingSymbol.cloneForAstAutomata());
    }
    return cloner;
  }

  @Override
  public int compareTo(BacktrackingBottomUpBranch that) {
    LinkedList<ReducingSymbol> thatReducingSymbols = that.reducingSymbols;
    // endIndexOfToken
    int thisEndIndexOfToken = this.reducingSymbols.getLast().endIndexOfToken;
    int thatEndIndexOfToken = thatReducingSymbols.getLast().endIndexOfToken;
    if (thisEndIndexOfToken != thatEndIndexOfToken) {
      return thisEndIndexOfToken - thatEndIndexOfToken;
    }
    // reducingSymbols
    if (this.reducingSymbols.size() != thatReducingSymbols.size()) {
      return this.reducingSymbols.size() - thatReducingSymbols.size();
    }
    ListIterator<ReducingSymbol> thisReducingSymbolsIt =
        this.reducingSymbols.listIterator(this.reducingSymbols.size());
    ListIterator<ReducingSymbol> thatReducingSymbolsIt =
        thatReducingSymbols.listIterator(thatReducingSymbols.size());
    while (thisReducingSymbolsIt.hasPrevious()) {
      ReducingSymbol thisReducingSymbol = thisReducingSymbolsIt.previous();
      ReducingSymbol thatReducingSymbol = thatReducingSymbolsIt.previous();
      int reducingSymbolCompare = thisReducingSymbol.compareTo(thatReducingSymbol);
      if (0 != reducingSymbolCompare) {
        return reducingSymbolCompare;
      }
    }
    return 0;
  }

  @Override
  public String toString() {
    StringBuilder stringBuilder = new StringBuilder();
    for (ReducingSymbol reducingSymbol : reducingSymbols) {
      stringBuilder
          .append("-")
          .append(reducingSymbol.astOfCurrentDfaState.grammar.name)
          .append(" ");
    }
    return stringBuilder.toString();
  }
}
